### PR TITLE
Add `ParametrizedLine`  and `fit_line`

### PR DESCRIPTION
**Pull request by edgarriba**
#### Changes

Implements `kornia.geometry.line` api based on Eigen.

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/geometry/linalg.py [hunk 1]
```diff
@@ -225,5 +226,10 @@ def point_line_distance(point: Tensor, line: Tensor, eps: float = 1e-9) -> Tenso
     return numerator / (denominator + eps)
 
 
+def squared_norm(x: Tensor) -> Tensor:
+    """Return the squared norm of a vector."""
+    return x.norm(2, -1) ** 2
```
Comment (lferraz) [inline kornia/geometry/linalg.py:231]: sum of x.^2 should be faster, right?

File: kornia/geometry/line.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+        l(t) = o + t * d
+    """
+
+    def __init__(self, origin: Tensor, direction: Tensor) -> None:
+        """Initializes a parametrized line of direction and origin.
+
+        Args:
+            origin: the origin point of the line of any dimension.
+            direction: the normalized vector direction of any dimension.
```
Comment (lferraz) [inline kornia/geometry/line.py:37]: why must it be normalized?

File: kornia/geometry/line.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    .. math::
+
+        l(t) = o + t * d
+    """
+
+    def __init__(self, origin: Tensor, direction: Tensor) -> None:
+        """Initializes a parametrized line of direction and origin.
+
+        Args:
+            origin: the origin point of the line of any dimension.
```
Comment (lferraz) [inline kornia/geometry/line.py:36]: there are infinite points maybe instead of 'the' should be 'any point on the line'

File: kornia/geometry/line.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        return self._origin
+
+    @property
+    def direction(self) -> Tensor:
+        """Return the line direction vector."""
+        return self._direction
+
+    def dim(self) -> int:
+        """Return the dimension in the line holds."""
+        return len(self.direction)
```
Comment (lferraz) [inline kornia/geometry/line.py:66]: should we check the number of dims of direction? here we are assuming that the dim is 1 but it could be any, right?

File: kornia/geometry/line.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    @classmethod
+    def through(cls, p0, p1) -> "ParametrizedLine":
+        """Constructs a parametrized line going from a point :math:`p0` to :math:`p1`.
+
+        Args:
+            p0: tensor with first point :math:`(B, D)`.
```
Comment (lferraz) [inline kornia/geometry/line.py:73]: i do not understand the meaining of B and D

File: kornia/geometry/line.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        diff: Tensor = point - self.origin
+        return squared_norm(diff - (self.direction @ diff) * self.direction)
+
+    def distance(self, point: Tensor) -> Tensor:
+        """Return the distance of a point to its projections onto the line.
+
+        Args:
+            point: the point to calculate the distance into the line.
+        """
+        return self.squared_distance(point).sqrt()
```
Comment (lferraz) [inline kornia/geometry/line.py:127]: squared_distance is quite innecifient used in this way. The funciton itself first compute the norm, then it is squared and now sqrt.... use directly the `norm` here.

Comment (lferraz) [inline kornia/geometry/line.py:127]: here the dot product + origin is not computed.

File: kornia/geometry/line.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        >>> weights = torch.ones(2, 10)
+        >>> direction = fit_line(points, weights)
+        >>> direction.shape
+        torch.Size([2, 3])
+    """
+    KORNIA_CHECK_IS_TENSOR(points, "points must be a tensor")
+    KORNIA_CHECK_SHAPE(points, ["B", "N", "D"])
+
+    points_mean = points.mean(-2, True)
+    A = points - points_mean
```
Comment (lferraz) [inline kornia/geometry/line.py:158]: I do not remember if this substaraction is mandatory but as far as I know you are changing teh line equation so some op should be done at the end. If you only return the "direction" then you are not fitting a line.

File: kornia/geometry/linalg.py (hunk 1 above)
Comment (edgarriba) [reply kornia/geometry/linalg.py:231]: done

File: kornia/geometry/line.py (hunk 3 above)
Comment (edgarriba) [reply kornia/geometry/line.py:36]: done

File: kornia/geometry/line.py (hunk 2 above)
Comment (edgarriba) [reply kornia/geometry/line.py:37]: i guess a design choice from Eigen

File: kornia/geometry/line.py (hunk 4 above)
Comment (edgarriba) [reply kornia/geometry/line.py:66]: https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Geometry/ParametrizedLine.h#L73

File: kornia/geometry/line.py (hunk 5 above)
Comment (edgarriba) [reply kornia/geometry/line.py:73]: D is the point number of dimensions

File: kornia/geometry/line.py (hunk 6 above)
Comment (edgarriba) [reply kornia/geometry/line.py:127]: can you specify more ?

File: kornia/geometry/line.py (hunk 7 above)
Comment (edgarriba) [reply kornia/geometry/line.py:158]: probably we should return the mean too

Comment (edgarriba) [reply kornia/geometry/line.py:158]: actually lets just return a parametrized line

File: kornia/geometry/line.py (hunk 2 above)
Comment (lferraz) [reply kornia/geometry/line.py:37]: the only reason I see is to compute distances. Do we want to force it?

File: kornia/geometry/line.py (hunk 5 above)
Comment (lferraz) [reply kornia/geometry/line.py:73]: and B is batch? Does parametrizedLines support batches?

File: kornia/geometry/line.py (hunk 4 above)
Comment (lferraz) [reply kornia/geometry/line.py:66]: but here `direction` is a tensor and you are not checking its dims

File: kornia/geometry/line.py (hunk 2 above)
Comment (edgarriba) [reply kornia/geometry/line.py:37]: i would sitck to Eigen o have compatibility. LAter we can make improvements based on user feedback

File: kornia/geometry/line.py (hunk 4 above)
Comment (edgarriba) [reply kornia/geometry/line.py:66]: ok, right. Missed that one  `len(self.direction.shape)`

File: kornia/geometry/line.py (hunk 5 above)
Comment (edgarriba) [reply kornia/geometry/line.py:73]: nope, we do. As soon as we don't implement any iterative method and we can vector is alright, which i think is the case for now

File: kornia/geometry/line.py (hunk 6 above)
Comment (edgarriba) [reply kornia/geometry/line.py:127]: `squared_distance` is different implementation of `squared_norm`
